validator/delphi_validator/validate.py

```diff
@@ -579,18 +579,149 @@ def check_rapid_change_num_rows(self, df_to_test, df_to_reference, checking_date
+        size_cut = 20
+        sig_cut = 3
+        sig_consec = 2.25
+
+        # A function mapped to each row to determine outliers based on fstat and ststat values 
+        def outlier_flag(frame):
+            if (abs(frame["val"]) > size_cut) and not (pd.isna(frame["ststat"])) and (frame["ststat"] > sig_cut):
+                return 1
+            if (abs(frame["val"]) > size_cut) and (pd.isna(frame["ststat"])) and not (pd.isna(frame["ftstat"])) and (frame["ftstat"] > sig_cut):
+                return 1
+            if (frame["val"] < -size_cut) and not (pd.isna(frame["ststat"])) and not (pd.isna(frame["ftstat"])):
+                return 1
+            return 0
+
+        # Calculate ftstat and ststat values for the rolling windows, group fames by geo region 
+        region_group = all_frames.groupby("geo_id")
+        window_size = 14
+        shift_val = 0
+        # Shift the window to match how R calculates rolling windows with even numbers
+        if (window_size%2 == 0):
+            shift_val = -1
+
+        all_full_frames = []
+        for cat, group in region_group:
+            rolling_windows = group["val"].rolling(window_size, min_periods=window_size)
+            center_windows = group["val"].rolling(window_size, min_periods=window_size, center=True)
+            fmean = rolling_windows.mean()
+            fmedian = rolling_windows.median()
+            smedian = center_windows.median().shift(shift_val)
+            fsd = rolling_windows.std()
+            ssd = center_windows.std().shift(shift_val)
+            vals_modified_f = group["val"] - fmedian.fillna(0)
+            vals_modified_s = group["val"] - smedian.fillna(0)
+            rolling_windows_f = vals_modified_f.rolling(window_size, min_periods=window_size)
+            center_windows_s = vals_modified_s.rolling(window_size, min_periods=window_size, center=True)
+            fmad = rolling_windows_f.median()
+            smad = center_windows_s.median().shift(shift_val)
+            ftstat = abs(vals_modified_f)/fsd
+            ststat = abs(vals_modified_s)/ssd
+            #print(vals_modified_f)
+            group['fmean'] = fmean
+            group['fmedian'] = fmedian
+            group['smedian'] = smedian
+            group['fsd'] = fsd
+            group['ssd'] = ssd
+            group['fmad'] = fmad
+            group['smad'] = smad
+            group['ftstat'] = ftstat
+            group['ststat'] = ststat 
+            all_full_frames.append(group) 
+
+        all_frames = pd.concat(all_full_frames)
+
+        # Determine outliers
+        outlier_source_df = all_frames.sort_values(by=['time_value']).copy()
+        outlier_source_df["flag"] = outlier_source_df.apply(outlier_flag, axis = 1)
+        outlier_group = outlier_source_df.groupby("geo_id")
+
+        outlier_append = pd.DataFrame()
+        for cat, group in outlier_group:
+            group = group.reset_index()
+            for index, row in group.iterrows():
```

Comment:
`iterrows()` is generally not recommended due to being slow. If you see a relatively easy way to vectorize this, that would be preferred. Otherwise, we can leave it and run profiling to see how slow it is compared to other bits of the code.